File: features/step_definitions/analizo_steps.rb
require 'fileutils'
require 'tmpdir'
require 'digest/sha1'

top_dir = FileUtils.pwd
saved_path = ENV["PATH"]
ENV['LC_ALL'] = 'C'

Before do
  ENV['PATH'] = top_dir + ':' + ENV['PATH']
  ENV["ANALIZO_CACHE"] = Dir.mktmpdir
end

After do
  FileUtils.rm_f('tmp.out')
  FileUtils.rm_f('tmp.err')
  FileUtils.rm_f(Dir.glob('*.tmp'))
  FileUtils.cd(top_dir)
  ENV['PATH'] = saved_path
  FileUtils.rm_rf(ENV["ANALIZO_CACHE"])
end

Given /^I am in (.+)$/ do |dir|
  FileUtils.cd(dir)
end

def get_tmpdir
  Dir.mktmpdir([Process.pid.to_s, '.analizo.tmpdir'])
end

Given /^I explode (.+)$/ do |tarball|
  tarball_full_path = File.expand_path(tarball)
  dirname = File.basename(tarball).sub('.tar.gz', '')
  tmpdir = get_tmpdir
  Dir.chdir(tmpdir) do
    system("tar xzf #{tarball_full_path}")
  end
  FileUtils.cd(File.join(tmpdir, dirname))
end

When /^I copy (.*) into a temporary directory$/ do |files|
  tmpdir = get_tmpdir
  FileUtils.cp_r(Dir.glob(files), tmpdir)
  FileUtils.cd(tmpdir)
end

When /^I change to an empty temporary directory$/ do
  FileUtils.cd(get_tmpdir)
end

def run(command)
  system("(#{command}) >tmp.out 2>tmp.err")
  if $?.is_a?(Fixnum)
    @exit_status = $?
  else
    @exit_status = $?.exitstatus
  end
  @stdout = File.readlines('tmp.out')
  @stderr = File.readlines('tmp.err')
end

When /^I run "([^\"]*)"$/ do |command|
  run command
end

at_exit do
  FileUtils.rm_rf(Dir.glob(File.join(Dir.tmpdir, '*.analizo.{out,err,tmpdir}')))
end

When /^I run "([^\"]*)" only once$/ do |command|
  run_marker = Digest::SHA1.hexdigest(Dir.pwd) + Digest::SHA1.hexdigest(command)
  stdout = File.join(Dir.tmpdir, run_marker + '.analizo.out')
  stderr = File.join(Dir.tmpdir, run_marker + '.analizo.err')
  if File.exist?(stdout) && File.exist?(stderr)
    @stdout = File.readlines(stdout)
    @stderr = File.readlines(stderr)
  else
    run command
    FileUtils.cp('tmp.out', stdout)
    FileUtils.cp('tmp.err', stderr)
  end
end

Then /^analizo must report that "([^\"]*)" depends on "([^\"]*)"$/ do |dependent, depended|
  dependent_regex = Regexp.escape dependent
  depended_regex = Regexp.escape depended
  if (@stdout.select { |line| line =~ /"#{dependent_regex}" -> "#{depended_regex}"/ }).size < 1
    raise AnalizoException.new("Output should say that %s depends on %s!" % [dependent.inspect, depended.inspect], @stdout, @stderr)
  end
end

Then /^the exit status must be (.+)$/ do |n|
  @exit_status.should == n.to_i
end

Then /^the exit status must not be (.+)$/ do |n|
  @exit_status.should_not == n.to_i
end

Then /^analizo must report that "([^\"]*)" is part of "([^\"]*)"$/ do |func,mod|
  line = (0...(@stdout.size)).find { |i| @stdout[i] =~ /subgraph "cluster_#{mod}"/ }
  func_regex = Regexp.escape func
  found = false
  if line
    for i in (line...(@stdout.size))
      if @stdout[i] =~ /^\s*\}\s*$/
        break
      elsif @stdout[i] =~ /node.*"#{func_regex}";/
        found = true
      end
    end
  end
  found.should == true
end

Then /^the output lines must match "([^\"]*)"$/ do |pattern|
  unless @stdout.join.match(pattern)
    raise AnalizoException.new("Output does not match %s! (expected to match)!" % pattern.inspect, @stdout, @stderr)
  end
end

Then /^the output must match "([^\"]*)"$/ do |pattern|
  unless @stdout.any? {|item| item.match(pattern) }
    raise AnalizoException.new("Output does not match %s! (expected to match)!" % pattern.inspect, @stdout, @stderr)
  end
end

Then /^the output must not match "([^\"]*)"$/ do |pattern|
  if @stdout.any? { |item| item.match(pattern) }
    raise AnalizoException.new("Output matches %s! (expected to NOT match)" % pattern.inspect, @stdout, @stderr)
  end
end

Then /^the contents of "(.+)" must match "([^\"]*)"$/ do |file, pattern|
  @out = File.readlines(file).join
  @out.should match(pattern)
end

Then /^analizo must emit a warning matching "([^\"]*)"$/ do |pattern|
  @stderr.join.should match(pattern)
end

module AnalizoRuby18Compatibiliy
  include Enumerable
  def each(&block)
    self.documents.each(&block)
  end
end

def __load_yaml_stream(data)
  stream = YAML.load_stream(data)
  if RUBY_VERSION < '1.9'
    stream.extend(AnalizoRuby18Compatibiliy)
  end
  stream
end

Then /^analizo must report that the project has (.+) = ([\d\.]+)$/ do |metric,n|
  __load_yaml_stream(@stdout.join).first[metric].should == n.to_f
end

Then /^analizo must report that module (.+) has (.+) = (.+)$/ do |mod, metric, value|
  module_metrics = __load_yaml_stream(@stdout.join).find { |doc| doc['_module'] == mod }
  case value
  when /^\d+|\d+\.\d+$/
    value = value.to_f
  when /^\[(.*)\]$/
    value = $1.split(/\s*,\s*/)
  end
  module_metrics[metric].should == value
end

Then /^analizo must present a list of metrics$/ do
  @stdout.size.should > 0

  @stdout.each do |line|
    line.should match(/(^[^-]+ - .+$)|(^Global Metrics:\n$)|(^Module Metrics:\n)|(^\n$)/)
  end
end

Given /^I create a file called (.+) with the following content$/ do |filename, table|
  File.open(filename, 'w') do |file|
    table.raw.each do |line|
      file.puts(line)
    end
  end
end

Then /^the file "([^\"]*)" should exist$/ do |file|
  File.exist?(file).should == true
end

Then /^the file "(.*?)" should have type (.*)$/ do |file, type|
  `file --brief --mime-type #{file}`.strip.should == type
end

Then /^analizo must present a list of languages$/ do
  @stdout.size.should > 0

  @stdout.each do |line|
    line.should match(/(^Languages:\n$)|(^\w+\n$)/)
  end
end

File: t/features/step_definitions/analizo_steps.pl
use strict;
use warnings;
use Test::More;
use Test::BDD::Cucumber::StepFile;
use File::Slurp;
use File::Temp qw( tempdir );
use File::Copy::Recursive qw( rcopy );
use YAML::XS;
use File::LibMagic;
use Archive::Extract;
use DBI;
use File::Spec;

our $exit_status;
our $stdout;
our $stderr;

use Env qw(@PATH $PWD);
push @PATH, "$PWD/blib/script", "$PWD/bin";

use IPC::Open3;
use Symbol 'gensym';

When qr/^I run "([^\"]*)"$/, sub {
  my ($c) = @_;
  my $command = $1;
  my ($IN, $STDOUT, $STDERR);
  $STDERR = gensym;
  my $pid = open3($IN, $STDOUT, $STDERR, "$command 2>tmp.err");
  waitpid $pid, 0;
  $exit_status = $?;
  local $/ = undef;
  $stdout = <$STDOUT>;
  $stderr = <$STDERR> . read_file('tmp.err');
};

When qr/^I run "([^\"]*)" on database "([^\"]*)"$/, sub {
  my ($c) = @_;
  my $statement = $1;
  my $db = $2;
  my @a = DBI->connect("dbi:SQLite:$db")->selectall_array($statement);
  $stdout = join("\n", map { join("|", @$_) } @a), "\n";
};

Then qr/^the output must match "([^\"]*)"$/, sub {
  my ($c) = @_;
  like($stdout, qr/$1|\Q$1\E/);
};

Then qr/^the output must not match "([^\"]*)"$/, sub {
  my ($c) = @_;
  unlike($stdout, qr/$1|\Q$1\E/);
};

Then qr/^the exit status must be (\d+)$/, sub {
  my ($c) = @_;
  cmp_ok($exit_status, '==', $1);
};

Then qr/^the exit status must not be (\d+)$/, sub {
  my ($c) = @_;
  cmp_ok($exit_status, '!=', $1);
};

Step qr/^I copy (.*) into a temporary directory$/, sub {
  my ($c) = @_;
  my $tmpdir = tempdir("analizo-XXXXXXXXXX", CLEANUP => 1, DIR => File::Spec->tmpdir);
  rcopy($1, $tmpdir);
  chdir $tmpdir;
};

Given qr/^I create a file called (.+) with the following content$/, sub {
  my ($c) = @_;
  open FILE, '>', $1 or die $!;
  if ($c->data =~ m/<\w+>/) {
    # The Test::BDD::Cucumber not support replace strings <key> in the content
    # by the values in table "Examples:", the code above does this.
    # Not yet know how find out what line of "Examples:" we are, then for now
    # we create entries for all values in the table.
    # TODO Implement it on Test::BDD::Cucumber in the right way and contribute
    # back to upstream.
    foreach my $row (@{ $c->scenario->data }) {
      foreach my $col (keys %$row) {
        (my $data = $c->data) =~ s/<$col>/$row->{$col}/sg;
        print FILE "$data";
      }
    }
  }
  else {
    print FILE $c->data;
  }
  close FILE;
};

Given qr/^I change to an empty temporary directory$/, sub {
  my ($c) = @_;
  chdir tempdir("analizo-XXXXXXXXXX", CLEANUP => 1, DIR => File::Spec->tmpdir);
};

Given qr/^I am in (.+)$/, sub {
  my ($c) = @_;
  chdir $1;
};

Then qr/^analizo must emit a warning matching "([^\"]*)"$/, sub {
  my ($c) = @_;
  like($stderr, qr/$1|\Q$1\E/);
};

Then qr/^analizo must not emit a warning matching "([^\"]*)"$/, sub {
  my ($c) = @_;
  unlike($stderr, qr/$1|\Q$1\E/);
};

Then qr/^analizo must report that "([^\"]*)" is part of "([^\"]*)"$/, sub {
  my ($c) = @_;
  my ($func, $mod) = ($1, $2);
  like($stdout, qr/subgraph "cluster_$mod" \{[^}]*node[^}]*"\Q$func\E";/);
};

Then qr/^analizo must report that "([^\"]*)" depends on "([^\"]*)"$/, sub {
  my ($c) = @_;
  my ($dependent, $depended) = ($1, $2);
  like($stdout, qr/"\Q$dependent\E" -> "\Q$depended\E"/);
};

Then qr/^the contents of "(.+)" must match "([^\"]*)"$/, sub {
  my ($c) = @_;
  my ($file, $pattern) = ($1, $2);
  like(read_file($file), qr/$pattern/);
};

Then qr/^analizo must report that the project has (.+) = ([\d\.]+)$/, sub {
  my ($c) = @_;
  my ($metric, $n) = ($1, $2);
  my @stream = Load($stdout);
  cmp_ok($stream[0]->{$metric}, '==', $n);
};

Then qr/^analizo must report that module (.+) has (.+) = (.+)$/, sub {
  my ($c) = @_;
  my ($module, $metric, $value) = ($1, $2, $3);
  my @stream = Load($stdout);
  my ($module_metrics) = grep { $_->{_module} && $_->{_module} eq $module } @stream;
  if ($module_metrics->{$metric}) {
    if ($value =~ /^\d+|\d+\.\d+$/) {
      cmp_ok($module_metrics->{$metric}, '==', $value);
    }
    elsif ($value =~ /^\[(.*)\]$/) {
      my @values = split(/\s*,\s*/, $1);
      is_deeply($module_metrics->{$metric}, \@values);
    }
  }
};

Then qr/^analizo must report that file (.+) not declares module (.+)$/, sub {
  my ($c) = @_;
  my ($filename, $module) = ($1, $2);
  my @stream = Load($stdout);
  my ($document) = grep { $_->{_module} && $_->{_module} eq $module } @stream;
  ok(!grep { /^$filename$/ } @{$document->{_filename}});
};

Then qr/^analizo must report that file (.+) declares module (.+)$/, sub {
  my ($c) = @_;
  my ($filename, $module) = ($1, $2);
  my @stream = Load($stdout);
  my ($document) = grep { $_->{_module} && $_->{_module} eq $module } @stream;
  ok(grep { /^$filename$/ } @{$document->{_filename}});
};

Then qr/^analizo must present a list of metrics$/, sub {
  my ($c) = @_;
  like($stdout, qr/Global Metrics:/);
  like($stdout, qr/Module Metrics:/);
};

Then qr/^analizo must present a list of languages$/, sub {
  my ($c) = @_;
  like($stdout, qr/Languages:/);
};

Then qr/^the file "([^\"]*)" should exist$/, sub {
  my ($c) = @_;
  my $file = $1;
  ok(-e $file);
};

Then qr/^the file "([^\"]*)" should not exist$/, sub {
  my ($c) = @_;
  my $file = $1;
  ok(! -e $file);
};

Then qr/^the file "(.*?)" should have type (.*)$/, sub {
  my ($c) = @_;
  my ($file, $type) = ($1, $2);
  my $magic = File::LibMagic->new;
  my $mime = $magic->checktype_filename($file);
  like($mime, qr/$type;/);
};

When qr/^I explode (.+)$/, sub {
  my ($c) = @_;
  my $tarball = $1;
  my $archive = Archive::Extract->new(archive => $tarball);
  $archive->extract(to => tempdir("analizo-XXXXXXXXXX", CLEANUP => 1, DIR => File::Spec->tmpdir));
  chdir $archive->extract_path;
};

Then qr/^the output lines must match "([^\"]*)"$/, sub {
  my ($c) = @_;
  my $pattern = $1;
  like($stdout, qr/$pattern/);
};
